src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -28,6 +28,56 @@ void GcsTaskManager::Stop() {
   }
 }
 
+bool GcsTaskManager::GcsTaskManagerStorage::ShouldMarkChildrenFailed(
+    const TaskID &child_task, const TaskID &parent_task, int64_t *failed_ts) const {
+  const auto &child_task_event = GetLatestTaskEvent(child_task);
+  if (!child_task_event.has_value()) {
+    return false;
+  }
+  // TODO(rickyx): Fix detached actor
+  // // A detached actor task should not be affected by its parent's tasks.
+  // if (child_task_event->has_task_info() &&
+  //     child_task_event->task_info().is_detached_actor()) {
+  //   return false;
+  // }
+
+  // If the child task has already finished/failed, we don't need to mark it
+  if (child_task_event->has_state_updates() &&
+      (child_task_event->state_updates().failed_ts() != 0 ||
+       child_task_event->state_updates().finished_ts() != 0)) {
+    return false;
+  }
+
+  if (parent_task.IsNil()) {
+    return false;
+  }
+
+  const auto &parent_task_event = GetLatestTaskEvent(parent_task);
+
+  if (!parent_task_event.has_value()) {
+    return false;
+  }
+
+  // Check if parent has failed.
+  if (!parent_task_event->has_state_updates()) {
+    return false;
+  }
+
+  auto state_updates = parent_task_event->state_updates();
+  if (state_updates.failed_ts() == 0) {
+    return false;
+  }
+
+  // Only if parent failed with worker died or actor died, we should mark the
+  // children tasks failed since the children tasks' workers are killed.
+  if (state_updates.error_info().error_type() == rpc::ErrorType::WORKER_DIED ||
```

Comment:
I think there are other cases this is not true. For example, NODE_DIED should be included.

What about we don't update only when it is the application failure? (and mark everything else as failed).

Reply:
good catch. let me tried to verify more explicitly. I wonder how task failures like placement group removed, local raylet died would behave. I will add this to the doc-string.